File: 1428-jump-game-iii/jump-game-iii.py
class Solution:
    def canReach(self, nums: List[int], start: int) -> bool:
        # bfs
        q = collections.deque()
        q.append(start)
        visited = set([start])

        while q:
            cur_idx = q.popleft()
            if nums[cur_idx] == 0:
                return True
            nbrs = [cur_idx - nums[cur_idx], cur_idx + nums[cur_idx]]
            for nbr in nbrs:
                if nbr not in visited and nbr >= 0 and nbr < len(nums):
                    q.append(nbr)
                    visited.add(nbr)

        return False
        

File: 1428-jump-game-iii/jump-game-iii.java
class Solution {
    public boolean canReach(int[] arr, int start) {
        // bfs
        // visited idxs
        // queue

        Deque<Integer> queue = new ArrayDeque<>();
        Set<Integer> visited = new HashSet<>();
        queue.offerLast(start);

        while (!queue.isEmpty()) {
            int cur = queue.pollFirst();
            if (arr[cur] == 0) {
                return true;
            }
            int[] nbrs = new int[] {cur - arr[cur], cur + arr[cur]};
            for (int nbr : nbrs) {
                if (nbr < 0 || nbr > arr.length - 1 || visited.contains(nbr)) {
                    continue;
                }
                visited.add(nbr);
                queue.offerLast(nbr);
            }
        }

        return false;
    }
}